### DISCUSSION TITLE
Screen frame calculation explanation

### DISCUSSION THREAD

**Discussion by danielstorch** [fingerprintjs/fingerprintjs#783] [Q&A] [answered]
So I searched for the values `screen.availWidth` and `screen.availHeight` and found that they are used in the component `screen_frame.ts`. But I don't quite understand the calculation of the values. For example, why is `screen.availWidth` and `screen.availLeft` subtracted from `screen.width`? 

I have looked through the issues/discussions where it was mentioned that `screen.availWidth` and `screen.availLeft` were not reliable enough. But the new calculation confuses me a bit. I still have the same problem where the values are different just because I set the browser to full screen on macOS.

Here are the two lines from `screen_frame.ts` that I don't understand:
```
replaceNaN(toFloat(s.width) - toFloat(s.availWidth) - replaceNaN(toFloat(s.availLeft), 0), null),
replaceNaN(toFloat(s.height) - toFloat(s.availHeight) - replaceNaN(toFloat(s.availTop), 0), null),
```

Maybe someone can enlighten me! Thanks!

**Comment by Valve** [comment]
@Finesse can you please take a look at the question above?

**Comment by Finesse** [comment, marked as answer]
Hi @danielstorch. The "available screen area" is an area where application windows may be located. The area is a rectangle. `availLeft`, `availTop` are the coordinates of the top left corner of the rectangle. `availWidth` and `availHeight` are the sizes of the rectangle. `width` and `height` are the whole screen size.  The available area matches the whole screen unless there no bars and docks on the screen. The following picture shows the sizes on a macOS desktop with a bar on the top and a dock on the bottom (it shows only the vertical sizes, the horizontal work the same way):

![screen sizes](https://user-images.githubusercontent.com/9006227/177951799-b38e5270-4dec-457b-93bf-eed729b921b8.svg)

The `screen_frame.ts` file calculates the size of the bars and the docks on all 4 screen sides. In order to calculate them, you need to find the distance between the available area and the screen edges, i.e. the size of the frame around the available area. The left and the top are straightforward: this is what `availLeft` and `availTop` are.

The bottom and the right are a bit harder. I'll explain the height case, the width case is similar. The equation is: `top bar height + available area height + bottom bar height = screen height`. You can solve it and find the bottom height: `bottom bar height = screen height - top bar height - available area height`. All the values on the right are stored in the `screen` object's properties.

We use `toFloat` because sometimes the values of the `screen` properties can be strings, and you can't perform math operations with string in JavaScript. We use `replaceNaN` because sometimes the string values can't be converted to numbers.

We use the screen frame size instead of the available rectangle position because usually the sizes of docks and bars don't change when the screen size changes. It makes the entropy source more stable. But unfortunately it is not 100% stable, [we know about it](https://github.com/fingerprintjs/fingerprintjs/issues/568#issuecomment-722272726). The frame size changes when you turn on/off fullscreen mode because there are not bars and docks in fullscreen. But the there are only 2 frame sizes in your case, so you can't make many identities by switching fullscreen mode.

The current set of entropy components in FingerprintJS provides a good balance between stability and sensitivity. If you want more stability and less sensitivity, turn off the "screen frame" component by following the [extending guide](https://github.com/fingerprintjs/fingerprintjs/blob/master/docs/extending.md).

Also, if you need higher accuracy with no compromises, consider trying FingerprintJS Pro which uses backend to smooth the fluctuations in the entropy components:

- [Use the free tier for FingerprintJS Pro](https://dashboard.fingerprint.com/signup), which allows making up to 20K identification events per month
- [Use the paid tier for FingerprintJS Pro](https://dashboard.fingerprint.com/signup), which allows to make unlimited number of identification events per month
